Mathlib/AlgebraicGeometry/EllipticCurve/Reduction.lean

```diff
@@ -0,0 +1,229 @@
+/-
+Copyright (c) 2025 Bryan Wang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bryan Wang
+-/
+import Mathlib.AlgebraicGeometry.EllipticCurve.VariableChange
+import Mathlib.RingTheory.DedekindDomain.AdicValuation
+import Mathlib.RingTheory.Valuation.Discrete.Basic
+
+/-!
+# Reduction of Weierstrass curves over local fields
+
+This file defines reduction of Weierstrass curves over local fields, or more generally,
+fraction fields of discrete valuation rings.
+
+## Main definitions
+
+* `IsIntegral`: a predicate expressing that a given Weierstrass equation
+  has integral coefficients.
+* `IsMinimal`: a predicate expressing that a given Weierstrass equation
+  has minimal valuation of discriminant among all isomorphic integral Weierstrass equations.
+* `reduction`: the reduction of a Weierstrass curve given by a minimal Weierstrass equation,
+  which is a Weierstrass curve over the residue field.
+* `IsGoodReduction`: a predicate expressing that a given minimal Weierstrass equation
+  has valuation of its discriminant equal to zero.
+
+## Main statements
+
+* `exists_integral`: any Weierstrass curve is isomorphic to one given by
+  an integral Weierstrass equation.
+* `exists_minimal`: any Weierstrass curve is isomorphic to one given by
+  a minimal Weierstrass equation.
+
+## References
+
+* [J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, weierstrass equation, minimal weierstrass equation, reduction
+-/
+
+namespace WeierstrassCurve
+
+variable (R : Type*) [CommRing R] [IsDomain R] [IsDiscreteValuationRing R]
+variable {K : Type*} [Field K] [Algebra R K] [IsFractionRing R K]
+
+section Minimal
+
+open IsDedekindDomain.HeightOneSpectrum
+open IsDiscreteValuationRing
+
+/-- A Weierstrass equation over the fraction field `K` is integral if
+it has coefficients in the DVR `R`. -/
+@[mk_iff]
+class IsIntegral (W : WeierstrassCurve K) : Prop where
+  integral : ∃ W_int : WeierstrassCurve R, W = W_int.baseChange K
+
+lemma isIntegral_of_val_le_one {W : WeierstrassCurve K}
+    (h₁ : valuation K (maximalIdeal R) W.a₁ ≤ 1)
+    (h₂ : valuation K (maximalIdeal R) W.a₂ ≤ 1)
+    (h₃ : valuation K (maximalIdeal R) W.a₃ ≤ 1)
+    (h₄ : valuation K (maximalIdeal R) W.a₄ ≤ 1)
+    (h₆ : valuation K (maximalIdeal R) W.a₆ ≤ 1) :
+    IsIntegral R W := by
+  use ⟨(exists_lift_of_le_one h₁).choose,
+      (exists_lift_of_le_one h₂).choose,
+      (exists_lift_of_le_one h₃).choose,
+      (exists_lift_of_le_one h₄).choose,
+      (exists_lift_of_le_one h₆).choose⟩
+  ext
+  all_goals
+    simp only [map_a₁, map_a₂, map_a₃, map_a₄, map_a₆]
+    apply (exists_lift_of_le_one _).choose_spec.symm
+    assumption
+
+theorem exists_integral (W : WeierstrassCurve K) :
```

Comment:
As far as I can tell, you could generalize this proof to work over an arbitrary valuation ring, not just a discrete one.